### ISSUE TITLE
Universal support for bfloat16 data, ie saving results of float32 operations as bfloat16 to memory by truncating

### ISSUE THREAD

**Issue by tarcey** [tensorflow/tensorflow#45255]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2 (google colab)
- Are you willing to contribute it (Yes/No): ?

**Describe the feature and the current behavior/state.**

bfloat16 is a truncated float32. That means that any float32 value can be cast to bfloat16 by simply discarding the high two bytes, for example by a 16bit-shift or mask and saving the remaining bits into a 2-byte type, or packing two truncated floats into a 4-byte type. Such operations are available universally. Currently, bfloat16 is only supported on TPUs and some CPUs, which makes sense because those also accelerate computation on bfloat16 values.

But all devices can benefit from bfloat16 data even when there is no half-precision computation support of any form: By doing calculations in float32, but truncating the results to 16bits ("casting" to bfloat16) before saving to memory, and expanding back to 32bits after loading from memory, one can improve memory utilization (allowing larger models) and also increase the effective memory bandwidth.
This would not utilize fast half-precision _processing_, but that is effectively unusable for training anyways when there is no fast mixed-precision support.

Because bfloat16 has the same range of values as float32 but at a lower precision, it can solve the problem of underruns/overflows arising when using pure float16 when training on devices that do not support mixed precision, while still providing the memory size and bandwidth benefits of 16-bit data.

Currently, setting the dtype policy to "bfloat16" like so:
`policy = tf.keras.mixed_precision.experimental.Policy('bfloat16')
tf.keras.mixed_precision.experimental.set_policy(policy)`
Does not raise errors immediately, but results in type errors when building the model:
`Value for attr 'T' of bfloat16 is not in the list of allowed values: double, float, half, complex64, complex128 [...]`
Some tensorflow operations and layers do not support bfloat16 on the Tesla T4 used in colab, notably LSTMs. Removing such layers allows building the model, but will raise another type error when attempting to train the model via model.fit().

**Will this change the current api? How?**
No change, support for bfloat16 already exists for certain devices. Implementing this feature could be done via creating another mixed-precision policy, like "data-bfloat16".

**Who will benefit from this feature?**
Everybody that uses devices without mixed precision or even any half precision support (for reference, only Volta, Turing, and Ampere GPUs support fast mixed precision. The still widely-used Pascal GPUs like the P100 don't support it), by allowing to take advantage of improved memory utilization and effective memory bandwidth without the disabling underrun issues when attempting to train using pure float16. CPUs that are often bandwidth-limited in vector operations might benefit the most from 16-bit data.

**Any Other info.**

This feature would not benefit anyone that uses the most cutting-edge hardware. But for reference, even the P100 used in google colab services would benefit from a feature like this. While it would make no difference to researchers and companies with access to expensive cutting-edge hardware, those make up only a part of tensorflow's users. Enabling everyone to experiment with models with up to twice the size given their current hardware would be an obvious benefit for most users of tensorflow.

**Comment by edervishaj** [comment]
I would be very interested in this feature!

**Comment by tilakrayal** [comment]
@tarcey,
There was a related PR(https://github.com/tensorflow/tensorflow/pull/59972) which was raised for the bfloat16 has been merged and also I tried to execute the sample code on tensorflow v2.15 (GPU) by importing the keras from the Keras 3 & the code was executed without any issue/error. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/fdc27fa1701f846a63b859b9dab4132c/untitled1575.ipynb). Thank you!
